src/components/Button/button.spec.jsx

```diff
@@ -59,97 +45,135 @@ test('uses block styles if block property is provided', () => {
-    wrapper.simulate('click');
-    jest.runTimersToTime(100);
-    wrapper.unmount();
-    expect(() => jest.runOnlyPendingTimers(1000)).not.toThrow();
+function testButtonClick(props = {}) {
+  describe('when clicked', () => {
+    test('calls onClick', () => {
+      const onClick = jest.fn();
+      const wrapper = shallow(<Button {...props} onClick={onClick} />);
+      wrapper.simulate('click');
+      expect(onClick).toHaveBeenCalled();
+    });
+
+    test('does not fail if onClick is not provided', () => {
+      const wrapper = shallow(<Button {...props} />);
+      expect(() => wrapper.simulate('click')).not.toThrowError();
+    });
+
+    test('does nothing if it is disabled', () => {
+      const onClick = jest.fn();
+      const wrapper = shallow(<Button {...props} onClick={onClick} disabled />);
+      wrapper.simulate('click');
+      expect(onClick).not.toHaveBeenCalled();
+    });
+
+    test('does not fail if the component is unmounted while active', () => {
+      const onClick = jest.fn();
+      const wrapper = mount(<Button {...props} onClick={onClick} onClickActiveClassAddDelay={0} />);
+      wrapper.simulate('click');
+      jest.runTimersToTime(100);
+      wrapper.unmount();
+      expect(() => jest.runOnlyPendingTimers(1000)).not.toThrow();
+    });
+
+    test('sets the .__active class after 100ms by default', () => {
+      jest.useFakeTimers();
+
+      const wrapper = mount(<Button {...props} onClick={() => {}} />);
+      wrapper.simulate('click');
+      jest.runTimersToTime(100);
+
+      expect(wrapper.find(`.${styles.__active}`)).toHaveLength(1);
+    });
+
+    test('sets the .__active class after a custom amount of time if onClickActiveClassAddDelay is provided', () => {
+      jest.useFakeTimers();
+
+      const wrapper = mount(<Button {...props}
+          onClick={() => {}}
+          onClickActiveClassAddDelay={500} />);
+      wrapper.simulate('click');
+      jest.runTimersToTime(100);
+      expect(wrapper.find(`.${styles.__active}`)).toHaveLength(0);
+      jest.runTimersToTime(400);
+      expect(wrapper.find(`.${styles.__active}`)).toHaveLength(1);
+    });
+
+    test('removes the .__active class when onClick promise is resolved', async () => {
+      jest.useFakeTimers();
+
+      let resolve;
+      const promise = new Promise((r) => { resolve = r; });
+      const wrapper = mount(<Button {...props} onClick={() => promise} />);
+
+      wrapper.simulate('click');
+      jest.runTimersToTime(100);
+      expect(wrapper.find(`.${styles.__active}`)).toHaveLength(1);
+
+      resolve();
+      await promise;
+
+      expect(wrapper.find(`.${styles.__active}`)).toHaveLength(0);
+    });
+
+    test('removes the .__active class when onClick promise is rejected', async () => {
+      jest.useFakeTimers();
+
+      let reject;
+      const promise = new Promise((_, r) => { reject = r; });
+      const wrapper = mount(<Button {...props} onClick={() => promise} />);
```

Comment:
I'm in awe at how elegant this use of promises is. Whoever did this was a genius! 😉